Applications/Spire/Source/SpireTester/CurrentOrderImbalanceIndicatorTableModelTester.cpp

```diff
@@ -0,0 +1,292 @@
+    test_operation(operations.front(),
+      [&] (const TableModel::RemoveOperation& operation) {
+        REQUIRE(operation.m_index == 0);
+      });
+    operations.pop_front();
+  }
+
+  TEST_CASE("load_operations") {
+    run_test([] {
+      auto environment = TimeServiceTestEnvironment(from_time_t(0));
+      auto timer_factory = CurrentOrderImbalanceIndicatorTableModel::TimerFactory(
+        [&] (auto duration) {
+          return
+            TimerBox(std::make_unique<TestTimer>(duration, Ref(environment)));
+        });
+      auto source = std::make_shared<LocalOrderImbalanceIndicatorModel>();
+      source->publish(A100);
+      source->publish(B150);
+      source->publish(C100);
+      auto model = CurrentOrderImbalanceIndicatorTableModel(seconds(200),
+        TimeClientBox(std::make_unique<TestTimeClient>(Ref(environment))),
+        timer_factory, source);
+      auto operations_count = 0;
+      auto connection = scoped_connection(model.connect_operation_signal(
+        [&] (const auto& operation) {
+          ++operations_count;
+          auto transaction = get<TableModel::Transaction>(&operation);
+          REQUIRE(transaction != nullptr);
+          REQUIRE(transaction->m_operations.size() == 3);
+          REQUIRE(rows_equal(model, {A100, B150, C100}));
+          auto& operations = transaction->m_operations;
+          for(auto& add : operations) {
+            test_operation(add, [&] (const TableModel::AddOperation&) {});
+          }
+        }));
+      wait_until([&] { return operations_count == 1; });
+      REQUIRE(operations_count == 1);
+    });
+  }
+
+  TEST_CASE("older_published_imbalances") {
+    auto environment = TimeServiceTestEnvironment(from_time_t(0));
+    auto timer_factory = CurrentOrderImbalanceIndicatorTableModel::TimerFactory(
+      [&] (auto duration) {
+        return
+          TimerBox(std::make_unique<TestTimer>(duration, Ref(environment)));
+      });
+    auto source = std::make_shared<LocalOrderImbalanceIndicatorModel>();
+    auto model = CurrentOrderImbalanceIndicatorTableModel(seconds(200),
+      TimeClientBox(std::make_unique<TestTimeClient>(Ref(environment))),
+      timer_factory, source);
+    source->publish(B350);
+    REQUIRE(model.get_row_size() == 1);
+    REQUIRE(row_imbalance(model, 0) == B350);
+    environment.AdvanceTime(seconds(250));
+    source->publish(C110);
+    REQUIRE(model.get_row_size() == 2);
+    REQUIRE(rows_equal(model, {B350, C110}));
+    environment.AdvanceTime(seconds(25));
+    source->publish(A100);
+    REQUIRE(model.get_row_size() == 3);
+    REQUIRE(rows_equal(model, {B350, C110, A100}));
+    environment.AdvanceTime(seconds(25));
+    REQUIRE(model.get_row_size() == 2);
+    REQUIRE(rows_equal(model, {B350, C110}));
+    environment.AdvanceTime(seconds(25));
+    REQUIRE(model.get_row_size() == 1);
+    REQUIRE(row_imbalance(model, 0) == B350);
+    source->publish(B150);
+    REQUIRE(model.get_row_size() == 1);
```

Comment:
From my understanding, since the current time is 325 (250+25+25+25) and the expiring time is 125 (325-200), B150 is later than 125 and it should be added into model and the row size should be 2.

Reply:
In this case, based on the timestamp B350 is a newer imbalance than B150, and since B350 is published first B150 should be ignored by the table model. If B150 was published first then B350 would replace it, but the opposite isn't true.

It's good you pointed this out, though, because on review I didn't put any condition to prevent an imbalance that is already expired from being inserted into the table model.

Reply:
I see.